ESQL: Enable expressions with arithmetic negation

alex-spies opened the pull request:
Evaluating ES|QL queries like
```
eval a = -1
| keep a
```
currently fails due to a missing `EvalMapper` for the expression corresponding to the arithmetic negation `-1`, the `NEG` expression.

Add this mapper to enable expression evaluations like this.

alex-spies committed:
Add mapper for NEG expressions

Evaluating ES|QL queries like `eval a = -1 | keep a` currently fails due
to a missing `EvalMapper` for the `NEG` expression, which corresponds to
the arithmetic negation `-1`.

Add this mapper to enable expression evaluations like this.

alex-spies committed:
Use Math.negateExact to avoid overflows

alex-spies committed:
Apply spotless

alex-spies committed:
Merge branch 'feature/esql' into fix/esql/neg-expression

alex-spies committed:
Merge branch 'feature/esql' into fix/esql/neg-expression

alex-spies committed:
Add tests for negating ints, longs and doubles

alex-spies committed:
Disallow negating unsigned longs

alex-spies committed:
Refactor unsigned negation validation

alex-spies committed:
Turn comment into doc string

Make comments on helper methods in Verifier.java consistent.

alex-spies committed:
Slightly refactor Verifier.java

alex-spies committed:
Apply spotless

alex-spies committed:
Merge branch 'feature/esql' into fix/esql/neg-expression

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

alex-spies commented:
Ready for review:
* Added tests for negating different datatypes.
* Added validation that forbids negating an unsigned long, as performing the negation only really makes sense if the unsigned long can be cast to long; implicitly casting to long, however, would be inconsistent with subtraction for longs, which requires the result to be a valid unsigned long again.
* Slightly cleaned up adjacent code by updating doc strings and visibility of some helper methods in `Verifier.java`.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -65,7 +66,13 @@ public Verifier(Metrics metrics) {
+     * @param plan The logical plan to be verified
+     * @return a collection of verification failures; empty if and only if the plan is valid
+     */
+    public Collection<Failure> verify(LogicalPlan plan) {
```
This method is not meant to be called publicly but instead only by the analyzer (or the tests).
I don't see it called anywhere else in the code either.
Keep the protected visibility.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/Neg.java:
```diff
@@ -0,0 +1,27 @@
+
+import org.elasticsearch.compute.ann.Evaluator;
+
+public class Neg {
```
Negate. We aim for clarity (also avoid a slang term).

alex-spies replied:
I prefer `Negate` as well! However, that would be inconsistent with the [corresponding expression](https://github.com/elastic/elasticsearch/blob/feature/esql/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/Neg.java) from `ql`. The evaluators defined inside the [esql expression subpackage](https://github.com/elastic/elasticsearch/tree/feature/esql/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/) seem to follow the structure of the [ql expression subpackage](https://github.com/elastic/elasticsearch/tree/feature/esql/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression) and this would break the pattern.

I could either
1. live with the inconsistency that the evaluator for `Neg` expressions is called `Negate`,
2. stay consistent, rename both to `Negate` (though that would create a bit of noise - but could be done in a follow-up PR), or
3. stick with `Neg`.

What do you think?

costin replied:
go with 3.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/Neg.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Evaluator(extraName = "Doubles")
+    static double processDoubles(double v) {
+        return -v;
```
Need to check for overflow negation here as well:
```
if (v == Double.MAX_VALUE || v == Double.MIN_VALUE) {
     throw new ArithmeticOverflow("double overflow");
}
```

bpintea replied:
I guess this should be fine, no? It's just a bit flip.

alex-spies replied:
That was my thinking as well. Anyway, I'll double check that and at least add a comment explaining why this is fine, if correct.

alex-spies replied:
Just manually tested negating `Double.MAX_VALUE` (and its negative), which is fine.

I don't see a good way to add a test case for this:
- wouldn't trust a csv test to correctly construct `Double.MAX_VALUE` since there may be rounding errors when parsing the literal)
- there's no existing test class where this would fit in, and it feels odd to add a new test class just for this.

So, if we're happy with that, I'd like to simply add a comment here that explains that negating a double always fine.

alex-spies replied:
Added unit tests for the Neg evaluators, also covering all edge cases.

costin reviewed (approved):
Left a few comments otherwise LGTM.
Needs a follow-up PR to handle in the grammar negating negations:
`Neg(Neg(Neg(Neg(ul))))" - should be allowed:
`row x =to_u(1) | eval a = ----x` or `eval a = -(-(-(-x))))`

See [ExpressionBuilder#minusAwareSource](https://github.com/elastic/elasticsearch/blob/54effe15be9790b5fd8ab9e5f30704f85443da5f/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java#L911) for the recipe for it. There's also https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java#L588

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/NegMapper.java:
```diff
@@ -0,0 +1,63 @@
+        if (type.isNumeric()) {
+            var childEvaluator = EvalMapper.toEvaluator(neg.field(), layout).get();
+
+            // Unsigned longs are unsupported by choice; negating them does not make much sense without converting to long;
```
Nit: I'd move this comment on top of LONG handling, which is "closer" to the UL type; and maybe also shorten it (IMO the reason ultimately being that conversion would require some implicit type change, less to do with subtraction consistency).
But not worth a new CI round just for this.

alex-spies replied:
Will happily move! Not sure which location you have in mind, though - a cursory search didn't turn up any locations with e.g. comments on long handling for me. Can you point me to where you would rather place this?

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/NegMapper.java:
```diff
@@ -0,0 +1,63 @@
+                return () -> doubles.apply(childEvaluator);
+            }
+        }
+        throw new AssertionError("arithmetic negation operator with unsupported data type");
```
Nit: might be worth logging the fraudulent type.

alex-spies replied:
To clarify: since the `AssertionError` seems to kill the server with the given error message, I'll add the data type to the error message - is that what you had in mind? (I didn't see usage of an actual logger in any of the other mapper classes.)

costin replied:
Further yet - never throw an error; an exception is enough.
AssertionError are used by unit tests.

alex-spies replied:
Agree!

Actually, `AssertionError` is used in multiple mappers (and I only used it to be consistent with the behavior of e.g. `ArithmeticMapper`.

To keep the PR from becoming noisy, how about I do a follow-up PR that kicks out all `AssertionError`s from the mappers and replaces them with an appropriate exception, @costin ?

alex-spies replied:
Went with @luigidellaquila 's suggestion of using `UnsupportedOperationException` for now, will remove `AssertionError` from other mappers in a follow-up PR.

bpintea reviewed (approved):
LGTM

alex-spies committed:
Revert visibility of Verifier::verify

alex-spies committed:
Enable (de-) serialization of Neg expression nodes

alex-spies committed:
Add Neg to EvalMapperTests

alex-spies committed:
Comment why negating a double never fails

alex-spies committed:
Log type if negation is unsupported

alex-spies commented:
> Needs a follow-up PR to handle in the grammar negating negations:
> `Neg(Neg(Neg(Neg(ul))))" - should be allowed: `row x =to_u(1) | eval a = ----x`or`eval a = -(-(-(-x))))`

To clarify: do we want `--x` to be allowed if `x` is of type unsigned long? Negating negations otherwise works fine and is also part of the [added test case](https://github.com/elastic/elasticsearch/blob/1e42c5fd6b591661095c499548fd3d74fea18eae/x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec#L125) - although I could add an expression that involves parentheses as well.

Of course we can support double negations for type `ulong`, but that's probably not a grammar issue, no? I'd expect that this instead requires amending the validation a bit + optimize away double negations.

costin commented:
> do we want --x to be allowed if x is of type unsigned long?

alex-spies committed:
Do not use String::format in error message

This API call is forbidden.

alex-spies committed:
Add multiple negation with parentheses to tests

alex-spies committed:
Move + shorten comment

alex-spies committed:
Fix typo

alex-spies committed:
Add unit tests for Neg evaluators

alex-spies committed:
Apply spotless

alex-spies committed:
Merge branch 'feature/esql' into fix/esql/neg-expression

alex-spies committed:
Merge branch 'feature/esql' into fix/esql/neg-expression

luigidellaquila reviewed (approved):
LGTM

About `AssertionError`s, I'd say let's not introduce new ones with this PR (eg. use UnsupportedOperationException) and then get rid of all the other occurrences with a separate PR

alex-spies committed:
Merge branch 'feature/esql' into fix/esql/neg-expression

alex-spies committed:
Exception not error on unsupported neg mapping

Mapping a Neg to an evaluator for unsupported types like unsigned long
should not result in  an AssertionError. Use
UnsupportedOperationException instead.
